src/modules/roc_netio/target_libuv/roc_netio/udp_sender_port.cpp

```diff
@@ -252,6 +265,33 @@ packet::PacketPtr UdpSenderPort::read_() {
     return pp;
 }
 
+bool UdpSenderPort::try_nonblocking_send_(const packet::PacketPtr& pp) {
+    if (pending_ == 0) {
+        const packet::UDP& udp = *pp->udp();
+        roc_log(LogTrace,
+                "udp sender: sending packet non-blocking: num=%u src=%s dst=%s sz=%ld",
+                packet_counter_ + 1,
+                address::socket_addr_to_str(config_.bind_address).c_str(),
+                address::socket_addr_to_str(udp.dst_addr).c_str(),
+                (long)pp->data().size());
+
+        const bool success =
+            sendto_nb(fd_, pp->data().data(), pp->data().size(), udp.dst_addr);
+        if (success) {
+            ++packet_counter_;
+            return true;
+        }
+        roc_log(LogTrace,
+                "udp sender: couldn't send packet non-blocking; falling back to "
+                "write_sem_: num=%u src=%s dst=%s sz=%ld",
+                packet_counter_ + 1,
+                address::socket_addr_to_str(config_.bind_address).c_str(),
+                address::socket_addr_to_str(udp.dst_addr).c_str(),
+                (long)pp->data().size());
```

Comment:
Nitpick: I think it would be easier to debug things if every packet had only one log message. Currently it can be 1 or 3 messages per packet.

I suggest to log in try_nonblocking_send_() only if the non-blocking write succeeded. This way, either try_nonblocking_send_() or write_sem_cb_() will log one message.